### ISSUE TITLE
Enable GCS logs

### ISSUE THREAD

**Issue by thockin** [kubernetes/k8s.io#904]
At least for gs://k8s-artifacts-prod, it would be nice to break down the details.

https://cloud.google.com/storage/docs/access-logs

**Comment by spiffxp** [comment]
Going to repurpose this as umbrella issue for:
- setting up group restricting access to these logs: https://github.com/kubernetes/k8s.io/pull/1871
- setting up IAM such that only this group has access to the logs (https://github.com/kubernetes/k8s.io/pull/1881 tried creating a custom role but I don't think this is necessary)
- enabling the logs

**Comment by spiffxp** [comment]
I'm guessing gs://k8s-artifacts-gcslogs is what was used when we trialed access logs and then disabled them
```console
$ gsutil ls -p k8s-artifacts-prod
gs://asia.artifacts.k8s-artifacts-prod.appspot.com/
gs://eu.artifacts.k8s-artifacts-prod.appspot.com/
gs://k8s-artifacts-cni/
gs://k8s-artifacts-cri-tools/
gs://k8s-artifacts-csi/
gs://k8s-artifacts-gcslogs/
gs://k8s-artifacts-kind/
gs://k8s-artifacts-prod/
gs://us.artifacts.k8s-artifacts-prod.appspot.com/
```

It's currently empty
```console
$ gsutil ls gs://k8s-artifacts-gcslogs
$
```

I can't find anything in git history that shows this bucket being created via our scripts

**Comment by spiffxp** [comment]
Let's look at what IAM is today

For the project containing the bucket
```yaml
$ gcloud projects get-iam-policy k8s-artifacts-prod --format="yaml(bindings)"
bindings:
- members:
  - serviceAccount:service-388270116193@compute-system.iam.gserviceaccount.com
  role: roles/compute.serviceAgent
- members:
  - serviceAccount:service-388270116193@container-analysis.iam.gserviceaccount.com
  role: roles/containeranalysis.ServiceAgent
- members:
  - group:k8s-infra-artifact-security@kubernetes.io
  - serviceAccount:k8s-infra-gcr-vuln-dashboard@k8s-artifacts-prod.iam.gserviceaccount.com
  role: roles/containeranalysis.occurrences.viewer
- members:
  - serviceAccount:service-388270116193@gcp-sa-containerscanning.iam.gserviceaccount.com
  role: roles/containerscanning.ServiceAgent
- members:
  - serviceAccount:388270116193-compute@developer.gserviceaccount.com
  - serviceAccount:388270116193@cloudservices.gserviceaccount.com
  - serviceAccount:service-388270116193@containerregistry.iam.gserviceaccount.com
  role: roles/editor
- members:
  - group:k8s-infra-artifact-admins@kubernetes.io
  role: roles/errorreporting.user
- members:
  - serviceAccount:service-388270116193@gcp-sa-pubsub.iam.gserviceaccount.com
  role: roles/iam.serviceAccountTokenCreator
- members:
  - group:k8s-infra-artifact-admins@kubernetes.io
  role: roles/run.admin
- members:
  - serviceAccount:k8s-infra-gcr-auditor-invoker@k8s-artifacts-prod.iam.gserviceaccount.com
  role: roles/run.invoker
- members:
  - serviceAccount:service-388270116193@serverless-robot-prod.iam.gserviceaccount.com
  role: roles/run.serviceAgent
- members:
  - group:k8s-infra-artifact-admins@kubernetes.io
  role: roles/serviceusage.serviceUsageConsumer
- members:
  - group:k8s-infra-artifact-admins@kubernetes.io
  - group:k8s-infra-artifact-security@kubernetes.io
  role: roles/viewer
```

For the bucket itself
```yaml
# $ gsutil iam get gs://k8s-artifacts-gcslogs | yq -y '{bindings}'
bindings:
  - members:
      - group:k8s-infra-artifact-admins@kubernetes.io
      - projectEditor:k8s-artifacts-prod
      - projectOwner:k8s-artifacts-prod
    role: roles/storage.legacyBucketOwner
  - members:
      - projectViewer:k8s-artifacts-prod
    role: roles/storage.legacyBucketReader
  - members:
      - group:cloud-storage-analytics@google.com
    role: roles/storage.legacyBucketWriter
  - members:
      - projectEditor:k8s-artifacts-prod
      - projectOwner:k8s-artifacts-prod
    role: roles/storage.legacyObjectOwner
  - members:
      - projectViewer:k8s-artifacts-prod
    role: roles/storage.legacyObjectReader
  - members:
      - group:cloud-storage-analytics@google.com
      - group:k8s-infra-artifact-admins@kubernetes.io
    role: roles/storage.objectAdmin
```

**Comment by spiffxp** [comment]
Since I keep forgetting what all the permissions are for each role
```sh
gsutil iam get gs://k8s-artifacts-gcslogs > bucket-iam.json
for r in $(<bucket-iam.json jq -r '.bindings | map(.role) | .[]'); do 
  gcloud iam roles describe $r --format='json(name,includedPermissions)'
done \
  | jq --slurp 'map({key: .name, value: .includedPermissions}) | from_entries' \
  > role-permissions.json

<bucket-iam.json yq \
  --yaml-output --indentless\
  --slurpfile a role-permissions.json \
  '$a[0] as $roles | .bindings |= map(. += {permissions: $roles[.role]}) | {bindings}'
```
```yaml
bindings:
- members:
  - group:k8s-infra-artifact-admins@kubernetes.io
  - projectEditor:k8s-artifacts-prod
  - projectOwner:k8s-artifacts-prod
  role: roles/storage.legacyBucketOwner
  permissions:
  - storage.buckets.get
  - storage.buckets.getIamPolicy
  - storage.buckets.setIamPolicy
  - storage.buckets.update
  - storage.objects.create
  - storage.objects.delete
  - storage.objects.list
- members:
  - projectViewer:k8s-artifacts-prod
  role: roles/storage.legacyBucketReader
  permissions:
  - storage.buckets.get
  - storage.objects.list
- members:
  - group:cloud-storage-analytics@google.com
  role: roles/storage.legacyBucketWriter
  permissions:
  - storage.buckets.get
  - storage.objects.create
  - storage.objects.delete
  - storage.objects.list
- members:
  - projectEditor:k8s-artifacts-prod
  - projectOwner:k8s-artifacts-prod
  role: roles/storage.legacyObjectOwner
  permissions:
  - storage.objects.get
  - storage.objects.getIamPolicy
  - storage.objects.setIamPolicy
  - storage.objects.update
- members:
  - projectViewer:k8s-artifacts-prod
  role: roles/storage.legacyObjectReader
  permissions:
  - storage.objects.get
- members:
  - group:cloud-storage-analytics@google.com
  - group:k8s-infra-artifact-admins@kubernetes.io
  role: roles/storage.objectAdmin
  permissions:
  - resourcemanager.projects.get
  - resourcemanager.projects.list
  - storage.objects.create
  - storage.objects.delete
  - storage.objects.get
  - storage.objects.getIamPolicy
  - storage.objects.list
  - storage.objects.setIamPolicy
  - storage.objects.update
```

**Comment by spiffxp** [comment]
I think the goal should be to give `k8s-infra-gcs-access-logs@kubernetes.io` `roles/storage.objectViewer` directly on `gs://k8s-artifacts-gcslogs`, and remove as many bucket-specific or project-level bindings that result in granting `storage.objects.get` (and probably `storage.objects.list`)

Removing or altering some of the project-level service agent bindings might be tricky, some ideas:
- put the bucket in an entirely separate project from k8s-artifacts-prod that is dedicated just for holding PII
- disable services if possible (do we need cloudrun and compute in k8s-artifacts-prod?)
- try using IAM conditions on project-level bindings to prevent access to the one bucket
- use custom roles without project-level `storage.objects.*` permissions and then grant roles/storage.objectViewer on the necessary GCS buckets

References
- https://cloud.google.com/storage/docs/access-logs
- https://cloud.google.com/storage/docs/access-control/iam#conditions
- https://cloud.google.com/storage/docs/access-control/iam-roles#basic-roles-modifiable
- https://cloud.google.com/storage/docs/access-control/iam-permissions
- https://cloud.google.com/storage/docs/access-control/iam-roles

**Comment by ameukam** [comment]
Audit logs have been activated and the logs are now stored in `k8s-artifacts-gcslogs`. 

@spiffxp Do we want to recreate `k8s-artifacts-gcslogs` with the right set of permissions or just bind the `roles/storage.objectViewer` to k8s-infra-gcs-access-logs@kubernetes.io ?

**Comment by bernokl** [comment]
I see I still do not have access to the logs.
```
gsutil cat gs://k8s-artifacts-gcslogs/us.artifacts.k8s-artifacts-prod.appspot.com_usage_2021_04_27_20_00_00_061005db711e07cd16_v0 | tail
AccessDeniedException: 403 HttpError accessing <https://storage.googleapis.com/storage/v1/b/k8s-artifacts-gcslogs/o/us.artifacts.k8s-artifacts-prod.appspot.com_usage_2021_04_27_20_00_00_061005db711e07cd16_v0?alt=media>: response: <{'x-guploader-uploadid': 'ABg5-UyHoHVvqM_s_vmSNxOTC6fAIxFwzC2PCCw7pqBlZ3DQfkq95upaPpdwH8GjbPI-yyecTPLQhxLZKu_51Yvh-4E', 'content-type': 'text/html; charset=UTF-8', 'date': 'Tue, 27 Apr 2021 23:10:39 GMT', 'vary': 'Origin, X-Origin', 'expires': 'Tue, 27 Apr 2021 23:10:39 GMT', 'cache-control': 'private, max-age=0', 'content-length': '87', 'server': 'UploadServer', 'alt-svc': 'h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"', 'status': '403'}>, content <bb@ii.coop does not have storage.objects.get access to the Google Cloud Storage object.>

```

**Comment by ameukam** [comment]
@bernokl You will get access once https://github.com/kubernetes/k8s.io/pull/1966 is merged and deployed.

**Comment by spiffxp** [comment]
>put the bucket in an entirely separate project from k8s-artifacts-prod that is dedicated just for holding PII

I will call this issue closed when we either:
- use a bucket in an entirely separate project
- remove all [basic role bindings](https://cloud.google.com/storage/docs/access-control/iam-roles#basic-roles-intrinsic) from the bucket to ensure no unintended access from `roles/owner`, `roles/editor` or `roles/viewer` trickling down from the resource hierarchy

**Comment by ameukam** [comment]
> Should I have access to the logs?
> 
> ```
> gsutil cat gs://k8s-artifacts-gcslogs/us.artifacts.k8s-artifacts-prod.appspot.com_usage_2021_04_27_20_00_00_061005db711e07cd16_v0 | tail
> 
> AccessDeniedException: 403 HttpError accessing 
> <https://storage.googleapis.com/storage/v1/b/k8s-artifacts-gcslogs/o/us.artifacts.k8s-artifacts-prod.appspot.com_usage_2021_04_27_20_00_00_061005db711e07cd16_v0?alt=media>: 
> response: <{'x-guploader-uploadid': 'ABg5-UyHoHVvqM_s_vmSNxOTC6fAIxFwzC2PCCw7pqBlZ3DQfkq95upaPpdwH8GjbPI-yyecTPLQhxLZKu_51Yvh-4E', 
> 'content-type': 'text/html; charset=UTF-8', 'date': 'Tue, 27 Apr 2021 , ......
> 'status': '403'}>, 
> 
> content <bb@ii.coop does not have storage.objects.get access to the Google Cloud Storage object.>
> ```

@bernokl you should be able to access it: 

```shell
gsutil iam get gs://k8s-artifacts-gcslogs
{
  "bindings": [
    {
      "members": [
        "group:k8s-infra-artifact-admins@kubernetes.io",
        "projectEditor:k8s-artifacts-prod",
        "projectOwner:k8s-artifacts-prod"
      ],
      "role": "roles/storage.legacyBucketOwner"
    },
    {
      "members": [
        "projectViewer:k8s-artifacts-prod"
      ],
      "role": "roles/storage.legacyBucketReader"
    },
    {
      "members": [
        "group:cloud-storage-analytics@google.com"
      ],
      "role": "roles/storage.legacyBucketWriter"
    },
    {
      "members": [
        "projectEditor:k8s-artifacts-prod",
        "projectOwner:k8s-artifacts-prod"
      ],
      "role": "roles/storage.legacyObjectOwner"
    },
    {
      "members": [
        "projectViewer:k8s-artifacts-prod"
      ],
      "role": "roles/storage.legacyObjectReader"
    },
    {
      "members": [
        "group:cloud-storage-analytics@google.com",
        "group:k8s-infra-artifact-admins@kubernetes.io"
      ],
      "role": "roles/storage.objectAdmin"
    },
    {
      "members": [
        "group:k8s-infra-gcs-access-logs@kubernetes.io"
      ],
      "role": "roles/storage.objectViewer"
    }
  ],
  "etag": "CAk="
}
```

**Comment by spiffxp** [comment]
Update on where we're at.

The following buckets have had access logs turned on
```yaml
# $ echo "logged_buckets:"; find audit/projects -name 'logging.json' | cut -d/ -f 5 | sort | sed -e 's/^/- /'
logged_buckets:
- asia.artifacts.k8s-artifacts-prod.appspot.com
- eu.artifacts.k8s-artifacts-prod.appspot.com
- k8s-artifacts-cni
- k8s-artifacts-cri-tools
- k8s-artifacts-csi
- k8s-artifacts-kind
- k8s-artifacts-prod
- us.artifacts.k8s-artifacts-prod.appspot.com
```

They currently log to `gs://k8s-artifacts-gcslogs` (hosted in `k8s-artifacts-prod`) with a prefix that matches the bucket name, e.g.
```yaml
# $ find audit/projects -name 'logging.json' | head -n1 | xargs cat | yq -y
logBucket: k8s-artifacts-gcslogs
logObjectPrefix: us.artifacts.k8s-artifacts-prod.appspot.co
```

I would like to see us do the following to close this out:
- enable logging for `k8s-staging-*` buckets
- send these logs to a bucket that is hosted in its own project, with access gated to those who are permitted to access PII (https://github.com/kubernetes/k8s.io/pull/2031 sets up the project and bucket)
- create a followup issue for what kind of public anonymized reporting we're going to do

**Comment by spiffxp** [comment]
/assign @hh 
please assign to whomever on your team is working on this

**Comment by spiffxp** [comment]
These permissions don't seem quite right. The new bucket seems to be missing a lot.

new
```yaml
# gsutil iam get gs://k8s-infra-artifacts-gcslogs | yq -y 'del(.etag)'
bindings:
  - members:
      - group:cloud-storage-analytics@google.com
    role: roles/storage.legacyBucketWriter
  - members:
      - group:cloud-storage-analytics@google.com
    role: roles/storage.objectAdmin
```

old 
```yaml
# gsutil iam get gs://k8s-artifacts-gcslogs | yq -y 'del(.etag)'
bindings:
  - members:
      - group:k8s-infra-artifact-admins@kubernetes.io
      - projectEditor:k8s-artifacts-prod
      - projectOwner:k8s-artifacts-prod
    role: roles/storage.legacyBucketOwner
  - members:
      - projectViewer:k8s-artifacts-prod
    role: roles/storage.legacyBucketReader
  - members:
      - group:cloud-storage-analytics@google.com
    role: roles/storage.legacyBucketWriter
  - members:
      - projectEditor:k8s-artifacts-prod
      - projectOwner:k8s-artifacts-prod
    role: roles/storage.legacyObjectOwner
  - members:
      - projectViewer:k8s-artifacts-prod
    role: roles/storage.legacyObjectReader
  - members:
      - group:cloud-storage-analytics@google.com
      - group:k8s-infra-artifact-admins@kubernetes.io
    role: roles/storage.objectAdmin
  - members:
      - group:k8s-infra-gcs-access-logs@kubernetes.io
    role: roles/storage.objectViewer
```

How did I notice this? I tried running rsync to copy files from the old bucket to the new bucket
```
$ gsutil -m rsync gs://k8s-artifacts-gcslogs gs://k8s-infra-artifacts-gcslogs
Building synchronization state...

At source listing 10000...
At destination listing 10000...
At source listing 20000...
At destination listing 20000...
# ...
At source listing 230000...
At destination listing 230000...
At destination listing 240000...
```

240K files is not nothing.  They're all in a flat namespace.  I'm beginning to wonder if we should instead try prefixing them with /'s per bucket, e.g. instead of:
```
gs://k8s-artifacts-gcslogs/asia.artifacts.k8s-artifacts-prod.appspot.com_storage_2021_04_09_07_00_00_00b72_v0
```
we could have
```
gs://k8s-artifacts-gcslogs/asia.artifacts.k8s-artifacts-prod.appspot.com/gcs_storage_2021_04_09_07_00_00_00b72_v0
```

such that a `gsutil ls gs://k8s-artifacts-gcslogs` returns a list of buckets that are being logged, not all of the logfiles themselves

**Comment by spiffxp** [comment]
I'm working https://github.com/kubernetes/k8s.io/issues/1968#issuecomment-876889232 to migrate gcs logs that were manually setup for prod buckets (mentioned / listed in https://github.com/kubernetes/k8s.io/issues/904#issuecomment-859932423)

**Comment by spiffxp** [comment]
With the completion of https://github.com/kubernetes/k8s.io/issues/1968 I am tempted to call this done.

But, I have concerns about the scalability/longevity of leaving a bucket with monotonically increasing data lying around:
- all files land at root, so `gsutil ls` can take... a long time
- there was ~2.2TB of data in the bucket last I synced
- there were O(400k) files

That said:
- There is no limit on the number of objects that can live in a bucket.
- I can't find that there's a limit on number of objects within a given prefix.

Questions:
- Should we do some kind of rollup of the hourly/daily files into files with longer time periods?
- Should we set a lifecycle policy to auto delete data after a certain amount of time?

**Comment by spiffxp** [comment]
Discussed during most recent meeting, outcome was:
- we're going to leave it as-is for now, the existing bucket has a 400d lifecycle
- doing this sustainably will be a blocker for putting the full pipeline of gcs logs -> bigquery -> reports into production
- at that time we'd like to consider things like:
  - making bigquery the source of truth
  - rolling up / aggregating older results over time

There now remain two bigquery datasets in k8s-artifacts-prod:
- `k8s-artifacts-prod:gcs_logs`
- `k8s-artifacts-prod:http_lb_logs`

We should delete the gcs_logs dataset.  @ameukam was going to followup with @justinsb about the `http_lb_logs` dataset

**Comment by spiffxp** [comment]
```
$ bq rm k8s-artifacts-prod:gcs_logs
rm: remove dataset 'k8s-artifacts-prod:gcs_logs'? (y/N) y
```

**Comment by ameukam** [comment]
We are done with this. Migration to Artifact Registry (https://github.com/kubernetes/k8s.io/issues/1343) will imply a different approach to get logs.

I'll also delete 
/close
